javatests/arcs/core/entity/HandleManagerTestBase.kt

```diff
@@ -0,0 +1,397 @@
+import org.junit.Ignore
+import org.junit.Test
+
+@Suppress("EXPERIMENTAL_API_USAGE")
+open class HandleManagerTestBase {
+    private val backingKey = RamDiskStorageKey("entities")
+
+    data class Person(
+        override val entityId: ReferenceId,
+        val name: String,
+        val age: Int,
+        val isCool: Boolean,
+        val bestFriend: Reference
+    ) : Entity {
+
+        override fun ensureIdentified(idGenerator: Generator, handleName: String) {}
+
+        override fun serialize() = RawEntity(
+            entityId,
+            singletons = mapOf(
+                "name" to name.toReferencable(),
+                "age" to age.toReferencable(),
+                "is_cool" to isCool.toReferencable(),
+                "best_friend" to bestFriend
+            ),
+            collections = emptyMap()
+        )
+
+        override fun reset() {
+            TODO("Not yet implemented")
+        }
+    }
+
+    val entity1 = Person("entity1", "Jason", 21, false, Reference("entity2", backingKey,null))
+    val entity2 = Person("entity2", "Jason", 22, true, Reference("entity1", backingKey, null))
+
+    val entitySpec = object : EntitySpec<Person> {
+        @Suppress("UNCHECKED_CAST")
+        override fun deserialize(data: RawEntity) = Person(
+            entityId = data.id,
+            name = (data.singletons["name"] as ReferencablePrimitive<String>).value,
+            age = (data.singletons["age"] as ReferencablePrimitive<Int>).value,
+            isCool = (data.singletons["is_cool"] as ReferencablePrimitive<Boolean>).value,
+            bestFriend = data.singletons["best_friend"]!!.toReference(backingKey, null)
+        )
+
+        override val SCHEMA = Schema(
+            listOf(SchemaName("Person")),
+            SchemaFields(
+                singletons = mapOf(
+                    "name" to FieldType.Text,
+                    "age" to FieldType.Number,
+                    "is_cool" to FieldType.Boolean,
+                    "best_friend" to FieldType.EntityRef("1234acf")
+                ),
+                collections = emptyMap()
+            ),
+            "1234acf"
+        )
+    }
+
+    private val singletonRefKey = RamDiskStorageKey("single-ent")
+    private val singletonKey = ReferenceModeStorageKey(
+        backingKey = backingKey,
+        storageKey = singletonRefKey
+    )
+
+    private val setRefKey = RamDiskStorageKey("set-ent")
+    private val setKey = ReferenceModeStorageKey(
+        backingKey = backingKey,
+        storageKey = setRefKey
+    )
+
+
+    lateinit var readHandleManager: EntityHandleManager
+    lateinit var writeHandleManager: EntityHandleManager
+
+    open var testRunner = { block: suspend CoroutineScope.() -> Unit ->
+        runBlockingTest { this.block() }
+    }
+
+    @Before
+    fun setup() {
+        RamDisk.clear()
+        DriverFactory.register(RamDiskDriverProvider())
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private suspend fun createWriteSingletonHandle(
```

Comment:
Totally optional, but I'd suggest moving all the helper methods to the bottom of the file.

Reply:
oh yes, I meant to do that!